fix(assets): make srcset parsing HTML spec compliant (#16323)

rarous opened the pull request:
fixes #16323

### Description

Simplifies the spitting of `srcset` image candidates with descriptors, so it is compliant with spec and not driven by every edge case somebody finds out is not working.

rarous committed:
fix: Make srcset parsing HTML spec compliant (#16323)

sapphi-red commented on packages/vite/src/node/utils.ts:
```diff
@@ -749,7 +763,7 @@ export function processSrcSet(
 ): Promise<string> {
   return Promise.all(
-    splitSrcSetDescriptor(srcs).map(async ({ url, descriptor }) => ({
+    parseSrcset(srcs).map(async ({ url, descriptor }) => ({
```
This function (`processSrcSet`) is used for replacing URLs in [CSS `image-set`](https://developer.mozilla.org/en-US/docs/Web/CSS/image/image-set).
https://github.com/vitejs/vite/blob/c0edd26bbfeb9a8d80ebaa420e54fbb7f165bd9b/packages/vite/src/node/plugins/css.ts#L1748
I think we need to split the function in to two, one for `srcset` attr (the new function you wrote) and one for CSS `image-set` (you can use the original function in this PR).

rarous replied:
Why? Is there different behaviour? AFAIK there is not. All tests are passing. I can add some test for the usage in `image-set`. If I keep the original function for `image-set` the same bug affecting `srcset` will be present in `image-set`. This fixed it in both.

The previous version was prone to errors because of over-specification. This implementation is loose, so it doesn't know all current possible CSS functions, but is open to future extensions of possible values. So there is no future need to change it when CSS spec introduces new image functions. The whole purpose in all usages is to find potential pairs of URI and descriptor.

rarous replied:
Will look at it

rarous replied:
I reworked the Regex and the logic for `ImageCandidate` parsing. There is nothing left from the original `srcset` code, so I removed the License mention.

The Regex has named capturing groups `url` and `descriptor`. The `url` group can be any CSS function, CSS string (single or double quoted) or URL string. The `descriptor` is anything after the space and before the comma, and can have optional `type(...)` for the `image-set` case.

sapphi-red replied:
Great! Thanks for handling the CSS `image-set` case as well 💚

sapphi-red commented on packages/vite/src/node/utils.ts:
```diff
@@ -720,26 +720,40 @@ interface ImageCandidate {
-    .map((s) => {
-      const src = s.replace(escapedSpaceCharacters, ' ').trim()
-      const url = imageSetUrlRE.exec(src)?.[0] ?? ''
+/**
```
I guess you used the code from `srcset`. In that case, let's add a license comment.
```suggestion
/*!
 * Based on https://github.com/sindresorhus/srcset
 * MIT License, Copyright (c) Sindre Sorhus <sindresorhus@gmail.com> (https://sindresorhus.com)
 */
/**
```

sapphi-red reviewed (commented):
Thanks!

rarous committed:
fix: make srcset parsing also conformant with CSS image-set spec

rarous commented:
HRM tests on Node 22 are a bit brittle :-/ The failure is unrelated to my changes and possible race condition. I got another error on my machine just in the first run of tests.

```
 FAIL  playground/hmr/__tests__/hmr.spec.ts > soft invalidate
AssertionError: expected 'soft-invalidation/index.js is transfo…' to match 'soft-invalidation/index.js is transfo…'

Expected: "soft-invalidation/index.js is transformed 2 times. child is now updated?"
Received: "soft-invalidation/index.js is transformed 3 times. child is now updated?"
```

rarous committed:
fix: srcset - rework regex to extract url and descriptor via group matches

rarous committed:
fix possibly undefined groups

sapphi-red commented on packages/vite/src/node/utils.ts:
```diff
@@ -720,36 +720,44 @@ interface ImageCandidate {
+       @see https://html.spec.whatwg.org/multipage/images.html#srcset-attribute
+       @see https://drafts.csswg.org/css-images-4/#image-set-notation
+       */
+      // eslint-disable-next-line regexp/no-super-linear-backtracking
+      /\s*(?<url>[\w-]+\([^)]*\)|"[^"]*"|'[^']*'|[^,]\S*[^,])\s*(?<descriptor>[^,]*(?:\stype\([^)]*\))?)\s*(?:,|$)/g,
```
```suggestion
      /(?:^|\s)(?<url>[\w-]+\([^)]*\)|"[^"]*"|'[^']*'|[^,]\S*[^,])\s*(?:\s(?<descriptor>[^,\s]+(?:\stype\([^)]*\))?)\s*)?(?:,|$)/g,
```
I think this lint can be fixed by replacing with this regex. The diffs are:

- replaced the first `\s*` with `(?:^|\s)`
- wrap descriptor part with `(?:\s` + `)?` and swapped `[^,]*` inside with `[^,\s]+`: require a single `\s` if a descriptor exists

rarous replied:
Thank you. Added your suggestions, extracted a variable and extended the description in the comment.

rarous replied:
There is an ordering problem, `type` can also be before `1x`

rarous replied:
Should be fixed.

rarous committed:
improved regex so we can remove eslint ignore rule, improved doc comment for regex

rarous committed:
Makes descriptor regex more generic, added test case for any order of descriptor parts

patak-cat reviewed (approved):
Nice!
